File: testdata/precision-suite/tp_import_idioms.js
// True positives for the ordinary Node import idioms.
//
// The corpus tested only `const child_process = require("child_process")` — the
// one shape where the local name equals the module name. A destructured require
// and an ESM import are how Node is normally written, and both were silently
// missed. Four of five idioms produced nothing before import resolution landed.
const { exec } = require("child_process");
const cp = require("child_process");

function destructuredRequire(req) {
  const user = req.query.cmd;
  exec("sh -c " + user); // nox-expect: TAINT-002
}

function aliasedRequire(req) {
  const user = req.query.cmd;
  cp.exec("sh -c " + user); // nox-expect: TAINT-002
}

File: testdata/precision-suite/tp_import_idioms.py
# True positives for the ordinary Python import idioms.
#
# The corpus tested only `import os` + `os.system(...)` — the ONE shape where
# the local name happens to equal the module name, which is the shape the
# catalog matched. `from os import system` and `import os as o` are how Python
# is normally written, and both were silently missed: the same flow, reported
# only when the import was written the unusual way.
#
# The suite scoring 1.00 recall while the engine missed the common idiom is why
# these samples exist. If a future change drops import resolution, recall here
# falls and says so.
#
# `request` arrives as a parameter, as in tp_injection.py, so the sample tests
# the import idiom of the SINK and nothing else.
from os import system
import os as operating_system
import subprocess as sp


def from_import(request):
    cmd = request.args.get("cmd")
    system("echo " + cmd)  # nox-expect: TAINT-002


def aliased_module(request):
    cmd = request.args.get("cmd")
    operating_system.system("echo " + cmd)  # nox-expect: TAINT-002


def aliased_package(request):
    cmd = request.args.get("cmd")
    sp.check_output("echo " + cmd, shell=True)  # nox-expect: TAINT-002
